# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/script/ScriptLoader.h
# Commit: f29451ea18dc
# Full Hash: f29451ea18dcf20e024c15c82bb814f9c2d9389e
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-02-18 09:08:22
# Regressor Bug: 1742437
# File Overlap Count: 1
# Description:
#   Bug 1742437 - Create DOM specific Module Loader; r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D137298
# ==============================================================================

diff -r c0bd1d29be62 -r f29451ea18dc dom/script/ScriptLoader.h
--- a/dom/script/ScriptLoader.h	Thu Feb 17 15:57:50 2022 +0000
+++ b/dom/script/ScriptLoader.h	Thu Feb 17 15:57:50 2022 +0000
@@ -23,10 +23,10 @@
 #include "mozilla/dom/LoadedScript.h"
 #include "mozilla/dom/JSExecutionContext.h"  // JSExecutionContext
 #include "mozilla/dom/ScriptLoadRequest.h"
+#include "ModuleLoaderBase.h"
 #include "mozilla/MaybeOneOf.h"
 #include "mozilla/MozPromise.h"
 #include "ScriptKind.h"
-#include "ModuleLoader.h"
 
 class nsCycleCollectionTraversalCallback;
 class nsIChannel;
@@ -56,8 +56,8 @@
 class AutoJSAPI;
 class DocGroup;
 class Document;
+class ModuleLoader;
 class LoadedScript;
-class ModuleLoader;
 class ScriptLoaderInterface;
 class ModuleLoadRequest;
 class ModuleScript;
@@ -132,8 +132,6 @@
    */
   void DropDocumentReference() { mDocument = nullptr; }
 
-  void EnsureModuleHooksInitialized() override;
-
   /**
    * Add an observer for all scripts loaded through this loader.
    *
@@ -470,11 +468,7 @@
   /**
    * Start a load for aRequest's URI.
    */
-  nsresult StartLoad(ScriptLoadRequest* aRequest) {
-    return aRequest->IsModuleRequest() ? StartModuleLoad(aRequest)
-                                       : StartClassicLoad(aRequest);
-  }
-
+  nsresult StartLoad(ScriptLoadRequest* aRequest);
   /**
    * Start a load for a classic script URI.
    * Sets up the necessary security flags before calling StartLoadInternal.
@@ -483,13 +477,6 @@
 
   /**
    * Start a load for a module script URI.
-   * Sets up the necessary security flags before calling StartLoadInternal.
-   * Short-circuits if the module is already being loaded.
-   */
-  nsresult StartModuleLoad(ScriptLoadRequest* aRequest) override;
-
-  /**
-   * Start a load for a module script URI.
    */
   nsresult StartLoadInternal(ScriptLoadRequest* aRequest,
                              nsSecurityFlags securityFlags);
@@ -547,7 +534,7 @@
 
   nsresult AttemptAsyncScriptCompile(ScriptLoadRequest* aRequest,
                                      bool* aCouldCompileOut);
-  nsresult ProcessRequest(ScriptLoadRequest* aRequest) override;
+  nsresult ProcessRequest(ScriptLoadRequest* aRequest);
   nsresult CompileOffThreadOrProcessRequest(ScriptLoadRequest* aRequest);
   void FireScriptAvailable(nsresult aResult, ScriptLoadRequest* aRequest);
   // TODO: Convert this to MOZ_CAN_RUN_SCRIPT (bug 1415230)
@@ -557,12 +544,6 @@
   // Implements https://html.spec.whatwg.org/#execute-the-script-block
   nsresult EvaluateScriptElement(ScriptLoadRequest* aRequest);
 
-  // Handles both offthread and synchronous modules
-  nsresult CompileOrFinishModuleScript(
-      JSContext* aCx, JS::Handle<JSObject*> aGlobal,
-      JS::CompileOptions& aOptions, ModuleLoadRequest* aRequest,
-      JS::MutableHandle<JSObject*> aModule) override;
-
   // Handles both bytecode and text source scripts; populates exec with a
   // compiled script
   nsresult CompileOrDecodeClassicScript(JSContext* aCx,
@@ -624,7 +605,6 @@
   nsresult PrepareLoadedRequest(ScriptLoadRequest* aRequest,
                                 nsIIncrementalStreamLoader* aLoader,
                                 nsresult aStatus);
-  void ProcessLoadedModuleTree(ModuleLoadRequest* aRequest) override;
 
   void AddDeferRequest(ScriptLoadRequest* aRequest);
   void AddAsyncRequest(ScriptLoadRequest* aRequest);
@@ -642,7 +622,7 @@
   // execution of the script.
   static bool ShouldCacheBytecode(ScriptLoadRequest* aRequest);
 
-  void RunScriptWhenSafe(ScriptLoadRequest* aRequest) override;
+  void RunScriptWhenSafe(ScriptLoadRequest* aRequest);
 
   /**
    *  Wait for any unused off thread compilations to finish and then
@@ -731,6 +711,52 @@
   RefPtr<ScriptLoader> mLoader;
 };
 
+class ModuleLoader final : public ModuleLoaderBase {
+ private:
+  virtual ~ModuleLoader();
+
+ public:
+  explicit ModuleLoader(ScriptLoader* aLoader);
+
+  ScriptLoader* GetScriptLoader() {
+    return static_cast<ScriptLoader*>(mLoader.get());
+  }
+
+  // Methods that must be overwritten by an extending class
+  void EnsureModuleHooksInitialized() override;
+
+  /**
+   * Start a load for a module script URI.
+   * Sets up the necessary security flags before calling StartLoadInternal.
+   * Short-circuits if the module is already being loaded.
+   */
+  nsresult StartModuleLoad(ScriptLoadRequest* aRequest) override;
+
+  void ProcessLoadedModuleTree(ModuleLoadRequest* aRequest) override;
+
+  nsresult CompileOrFinishModuleScript(
+      JSContext* aCx, JS::Handle<JSObject*> aGlobal,
+      JS::CompileOptions& aOptions, ModuleLoadRequest* aRequest,
+      JS::MutableHandle<JSObject*> aModuleScript) override;
+
+  // Create a top-level module load request.
+  static already_AddRefed<ModuleLoadRequest> CreateTopLevel(
+      nsIURI* aURI, ScriptFetchOptions* aFetchOptions,
+      const SRIMetadata& aIntegrity, nsIURI* aReferrer, ScriptLoader* aLoader,
+      DOMScriptLoadContext* aContext);
+
+  // Create a module load request for a static module import.
+  already_AddRefed<ModuleLoadRequest> CreateStaticImport(
+      nsIURI* aURI, ModuleLoadRequest* aParent) override;
+
+  // Create a module load request for dynamic module import.
+  static already_AddRefed<ModuleLoadRequest> CreateDynamicImport(
+      nsIURI* aURI, ScriptFetchOptions* aFetchOptions, nsIURI* aBaseURL,
+      DOMScriptLoadContext* aContext, ScriptLoader* aLoader,
+      JS::Handle<JS::Value> aReferencingPrivate,
+      JS::Handle<JSString*> aSpecifier, JS::Handle<JSObject*> aPromise);
+};
+
 }  // namespace dom
 }  // namespace mozilla
 